### PR TITLE
Import hardware data and certificates from C3

### PR DESCRIPTION
**Pull request by nadzyah**
This PR resolves [C3-648](https://warthogs.atlassian.net/browse/C3-648)

This PR adds the logic for importing the following objects from C3 using `/api/v2/publiccertificates` API endpoint (https://github.com/canonical/hexr/pull/293):
* Vendors
* Platforms
* Configurations
* Machines (PhysicalMachines)
* Certificates
* Reports
* Release
* Bios

The import can be done by executing the `import_from_c3.py` script, what is also done during the container build

[C3-648]: https://warthogs.atlassian.net/browse/C3-648?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: server/README.md [hunk 1]
```diff
@@ -1,6 +1,24 @@
 # hwapi server
 
 
+## Importing Data from C3
```
Comment (mz2) [inline server/README.md:4]: Hmm, I think this could be rather just done at image build time?

Comment (mz2) [inline server/README.md:4]: I think this importing would be best done at image build time (a tool separate from the service would collate the data into a .sqlite, and that would then be copied into the image).

**Review by mz2 — CHANGES_REQUESTED** [review]
Someone else should review the implementation for correctness, but I think at a high level it would be preferable if the database were completely immutable and the importing were done at build time (by a separate tool that is used as a prerequisite to `docker build`).

File: server/hwapi/external/c3/api.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                    kernel=kernel,
+                    bios=bios,
+                    certificate=certificate,
+                )
+            except IntegrityError:
+                logging.error(
+                    "Got an error while importing certificates", exc_info=True
+                )
+                self.db.rollback()
+                continue
```
Comment (omar-selo) [inline server/hwapi/external/c3/api.py:151]: what does the rollback do here? given that there is a continue after it. Would it rollback all objects added so far even ones that had no errors?

File: server/hwapi/data_models/models.py [hunk 3]
```diff
@@ -74,7 +78,7 @@ class Machine(Base):
 class Certificate(Base):
     __tablename__ = "certificate"
     created_at: Mapped[datetime] = mapped_column(nullable=False)
-    name: Mapped[str] = mapped_column(String(80), nullable=True)
+    name: Mapped[str] = mapped_column(String(80), unique=True, nullable=True)
```
Comment (omar-selo) [inline server/hwapi/data_models/models.py:81]: `Certificate.name` is the name of what? I'm assuming certificates don't have name. If it's the name of a machine, then perhaps we should rename this variable to `machine_name`. But at the same time, I'm not sure what use we have with machine name

**Review by omar-selo — CHANGES_REQUESTED** [review]
Looks good, just a few comments

File: server/hwapi/data_models/models.py (hunk 3 above)
Comment (nadzyah) [reply server/hwapi/data_models/models.py:81]: This is what we call a certificate name, it's like a canonical ID but for certificates

image

Comment (omar-selo) [reply server/hwapi/data_models/models.py:81]: Oh I see. Do we need this bit of info in `hardware-api`?

File: server/hwapi/data_models/data_validators/software.py [hunk 4]
```diff
@@ -22,14 +22,13 @@
 
 
 class KernelPackageValidator(BaseModel):
-    name: str
+    name: str | None
```
[the remark this answers is not in the document]
Comment (nadzyah) [reply server/hwapi/data_models/data_validators/software.py:25]: the problem is that we don't have this data in C3. We have only kernel version:
https://github.com/canonical/hexr/blob/4f9474587fad7de77d92307fd5424e5701d1ce72/apps/hardware/models.py#L972-L973

File: server/schemas/openapi.yaml [hunk 5]
```diff
@@ -16,24 +16,24 @@ components:
       type: object
     BiosValidator:
       properties:
-        firmware_revision:
-          title: Firmware Revision
-          type: string
         release_date:
-          format: date
+          anyOf:
+          - format: date
```
[the remark this answers is not in the document]
Comment (nadzyah) [reply server/schemas/openapi.yaml:21]: but server also returns date for this field

File: server/Dockerfile [hunk 6]
```diff
@@ -12,4 +12,12 @@ RUN pip3 install poetry && \
 
 COPY . .
 
+ARG DB_URL
+ENV DB_URL=$DB_URL
+ARG C3_URL
+ENV C3_URL=$C3_URL
+
+ARG DATA_LOADING_SCRIPT_FILE=import_from_c3.py
+RUN if [ -n "$DATA_LOADING_SCRIPT_FILE" ]; then python scripts/$DATA_LOADING_SCRIPT_FILE; fi
```
[the remark this answers is not in the document]
Comment (nadzyah) [reply server/Dockerfile]: Can you please try once again, I've updated the readme instructions?

[the remark this answers is not in the document]
Comment (nadzyah) [reply server/Dockerfile]: I've updated it to use script path instead. Also, I'm using `-x` instead c:

File: server/hwapi/external/c3/api.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        """
+        logger.info(f"Importing certificates and machines from {urls.C3_URL}")
+        response = requests.get(
+            urls.CERTIFIED_CONFIGURATIONS_URL + urls.LIMIT_OFFSET, timeout=120
+        )
+        response.raise_for_status()
+        objects = response.json()["results"]
+        print(objects)
+        for obj in objects:
+            try:
```
[the remark this answers is not in the document]
Comment (nadzyah) [reply server/hwapi/external/c3/api.py]: I've moved this logic to another function. The tests for this logic are located under the `tests/external/c3` directory

**Review by omar-selo — APPROVED** [review]
Looks good to me

File: server/docker-compose.yml [hunk 8]
```diff
@@ -4,42 +4,42 @@ services:
...
       - "8080:8080"
-    environment:
-      - DB_URL=sqlite:////home/app/data/hwapi-dev.db
 
   hwapi-test:
     build:
       context: .
       dockerfile: Dockerfile
+      args:
+        DATA_LOADING_SCRIPT_FILE:
```
[the remark this answers is not in the document]
Comment (plars) [reply server/docker-compose.yml]: Maybe I'm missing something? But here's what I'm doing. Per the README.md, I'm running this for the tests:
`docker-compose up --attach-dependencies --force-recreate --abort-on-container-exit --build hwapi-test`

Here's what I'm getting back if I don't have DATA_LOADING_SCRIPT_FILE: ""
```
...
 => ERROR [hwapi-test 7/7] RUN if [ -n "import_from_c3.py" ]; then python  1.0s
------                                                                          
 > [hwapi-test 7/7] RUN if [ -n "import_from_c3.py" ]; then python scripts/import_from_c3.py; fi:                                                               
0.890 INFO:__main__:Importing data from C3
0.890 INFO:hwapi.external.c3.api:Importing certified configurations and machines from https://c3_url
0.907 Traceback (most recent call last):
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connection.py", line 198, in _new_conn
0.907     sock = connection.create_connection(
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/util/connection.py", line 60, in create_connection
0.907     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
0.907   File "/usr/local/lib/python3.10/socket.py", line 955, in getaddrinfo
0.907     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
0.907 socket.gaierror: [Errno -2] Name or service not known
0.907 
0.907 The above exception was the direct cause of the following exception:
0.907 
0.907 Traceback (most recent call last):
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 793, in urlopen
0.907     response = self._make_request(
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 491, in _make_request
0.907     raise new_e
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 467, in _make_request
0.907     self._validate_conn(conn)
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 1099, in _validate_conn
0.907     conn.connect()
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connection.py", line 616, in connect
0.907     self.sock = sock = self._new_conn()
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connection.py", line 205, in _new_conn
0.907     raise NameResolutionError(self.host, self, e) from e
0.907 urllib3.exceptions.NameResolutionError: <urllib3.connection.HTTPSConnection object at 0x758eedbb1600>: Failed to resolve 'c3_url' ([Errno -2] Name or service not known)
0.907 
0.907 The above exception was the direct cause of the following exception:
0.907 
0.907 Traceback (most recent call last):
0.907   File "/usr/local/lib/python3.10/site-packages/requests/adapters.py", line 486, in send
0.907     resp = conn.urlopen(
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 847, in urlopen
0.907     retries = retries.increment(
0.907   File "/usr/local/lib/python3.10/site-packages/urllib3/util/retry.py", line 515, in increment
0.908     raise MaxRetryError(_pool, url, reason) from reason  # type: ignore[arg-type]
0.908 urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='c3_url', port=443): Max retries exceeded with url: /api/v2/public-certificates/?pagination=limitoffset&limit=0 (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x758eedbb1600>: Failed to resolve 'c3_url' ([Errno -2] Name or service not known)"))
0.908 
0.908 During handling of the above exception, another exception occurred:
0.908 
0.908 Traceback (most recent call last):
0.908   File "/home/app/scripts/import_from_c3.py", line 40, in <module>
0.908     c3_api.load_certified_configurations()
0.908   File "/home/app/hwapi/external/c3/api.py", line 48, in load_certified_configurations
0.908     response = requests.get(
0.908   File "/usr/local/lib/python3.10/site-packages/requests/api.py", line 73, in get
0.908     return request("get", url, params=params, **kwargs)
0.908   File "/usr/local/lib/python3.10/site-packages/requests/api.py", line 59, in request
0.908     return session.request(method=method, url=url, **kwargs)
0.908   File "/usr/local/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
0.908     resp = self.send(prep, **send_kwargs)
0.908   File "/usr/local/lib/python3.10/site-packages/requests/sessions.py", line 703, in send
0.908     r = adapter.send(request, **kwargs)
0.908   File "/usr/local/lib/python3.10/site-packages/requests/adapters.py", line 519, in send
0.908     raise ConnectionError(e, request=request)
0.908 requests.exceptions.ConnectionError: HTTPSConnectionPool(host='c3_url', port=443): Max retries exceeded with url: /api/v2/public-certificates/?pagination=limitoffset&limit=0 (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x758eedbb1600>: Failed to resolve 'c3_url' ([Errno -2] Name or service not known)"))
------
failed to solve: process "/bin/sh -c if [ -n \"$DATA_LOADING_SCRIPT_FILE\" ]; then python scripts/$DATA_LOADING_SCRIPT_FILE; fi" did not complete successfully: exit code: 1
```

[the remark this answers is not in the document]
Comment (plars) [reply server/docker-compose.yml]: looks like there were changes since I tried it and posted that, it's working fine for me now. Thanks!

**Review by plars — APPROVED** [review]
my concerns were fixed, great work! +1 from me

File: README.md [hunk 9]
```diff
@@ -1,28 +1,76 @@
...
+### Using pre-populated database
+
+You can use this option if you already have the SQLite DB file and want to use it. Make sure that the DB
+is located under the `./server/` directory.
+
+You can stand up the environment by running the following commands:
+
+```bash
+# Assuming that the path to the DB file is ./hwapi.db
+docker-compose build --build-arg IMPORT_TOOL_PATH="" --build-arg DB_URL=sqlite:///./hwapi.db hwapi-dev
```
Comment (mz2) [inline README.md:30]: Hmm, does the Dockerfile image even need the `IMPORT_TOOL_PATH` arg at all if the import tool can just be run outside on the host / in a VM or container started up by a CI system?

Comment (nadzyah) [reply README.md:30]: I assumed importing data is the default behaviour, so you can run `docker-compose up` and get the container with the populated data. I thought that this is what we also want to have by default in production, that’s why I’ve set the default value, not an empty string
Do you think the default behaviour should be not to populate the DB at all?

Comment (mz2) [reply README.md:30]: If the Dockerfile can be kept simple, I'd keep it simple. 1) You have the import tool and 2) you have the Dockerfile that can use the DB, no matter how sourced. This is composable, and I don't think you need to specifically need to provide an option for the image build to execute that script.

Comment (mz2) [reply README.md:30]: For `docker-compose up` to work for a fresh checkout, IMO best option would be a small example data with some toy data.

File: server/schemas/openapi.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (mz2) [reply server/schemas/openapi.yaml:26]: I may be misunderstanding OpenAPI here: the format should be a date, it should never be a string.

[the remark this answers is not in the document]
Comment (nadzyah) [reply server/schemas/openapi.yaml:26]: But it’s not a string, it’s either date or null

**Review by mz2 — COMMENTED** [review]
I'd still the import tool path to be entirely dropped, but let's talk tomorrow in case you want to approach it differently.

**Review by mz2 — APPROVED** [review]
We've discussed this some more with Nadzeya, I'll propose a follow-up to turn this into a multi-stage build later.

**Pull request merged**